bpf/bpf_overlay.c

```diff
@@ -458,23 +468,43 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx,
 	}
 #endif /* ENABLE_EGRESS_GATEWAY_COMMON */
 
-	/* Deliver to local (non-host) endpoint: */
-	ep = lookup_ip4_endpoint(ip4);
-	if (ep && !(ep->flags & ENDPOINT_MASK_HOST_DELIVERY))
-		return ipv4_local_delivery(ctx, ETH_HLEN, *identity, MARK_MAGIC_IDENTITY,
-					   ip4, ep, METRIC_INGRESS, false, true, 0);
+#if defined(ENABLE_DSR) && defined(IS_BPF_OVERLAY) && (DSR_ENCAP_MODE == DSR_ENCAP_GENEVE)
+	/* Pass packets which will be returned using Geneve DSR
+	 * to host-stack for conntrack entry insertion.
+	 * This is needed because the packet will be masqueraded
+	 * by iptables if the conntrack entry isn't exist.
+	 */
+	if (!is_defined(ENABLE_MASQUERADE_IPV4)) {
```

Comment:
To clarify - what we want is netfilter symmetry between the forward and reply path, correct? In the reply path, netfilter will observe the traffic unless `ENABLE_HOST_ROUTING` is set. I would have expected us to match this behavior in the forward path (rather than checking for `ENABLE_MASQUERADE_IPV4`), same as we do [here](https://github.com/cilium/cilium/blob/eaecc8e9bf04389f8f328750762d7c6fe2b3b721/bpf/bpf_host.c#L742) for DSR traffic with IP-option.